### DISCUSSION TITLE
Inizializzare una variabile all'interno di un'automazione

### DISCUSSION THREAD

**Discussion by Newreader** [indomus/forum#1989] [GENERALE - Discussioni generiche]
Salve, ho la necessità di utilizzare un variabile da incrementare con valori numerici predefiniti all'interno di una automazione.
L'istruzione che sto utilizzando è la seguente:
{%- set contatore = 0 -%}
indentata secondo la posizione che occupa l'istruzione scritta nella riga superiore.
Per incrementarla:
{%- set contatore = contatore + 8 -%}
o altri valori che vengono determinati al verificarsi di determinate condizioni
Quando inserisco la prima istruzione (come pure la seconda) negli Strumenti per Sviluppatori -> Modelli va tutto bene, invece inserita nell'automazione mi restituisce il seguente errore:
Error loading /config/configuration.yaml: while scanning for the next token
found character '%' that cannot start any token
in "/config/automations.yaml", line 30, column 6
La parte iniziale dell'automazione è la seguente:
```- id: superamento_max_potenza
  alias: "Superamento Max Potenza"
  mode: single
  trigger:
    - platform: numeric_state
      entity_id: sensor.potenza_a2
      above: 2900
  condition: []
  action:
    {%- set contatore = 0 -%}
``` 
Grazie

**Comment by 50494554524F** [comment]
il codice tra  {% %} è codice jinja, lo puoi usare solo dove vengono definiti i template...

nel tuo caso ti serve un input_number e relativi servizi, o puoi usare variable+history da installare con hacs che ti permette di creare anche  variabili al volo... se, invece, non devi conservare i valori del contatore puoi usare "variables:" (leggi la documentazione sugli script)...

**Reply by Scialla** [reply]
Più che altro credo dovrebbe configurare un service prima di sparare codice Jinja

**Comment by Newreader** [comment]
Grazie per le risposte. Ieri avevo iniziato ad utilizzare "variables", ma il suo più grande limite nel mio caso è che le variabili così create hanno ambito locale con tutto ciò che ne consegue.
Comunque spiego, nella mia logica, a cosa mi serve la variabile diciamo "contatore" che voglio utilizzare.
Ho quattro prese intelligenti che mi forniscono il valore della potenza rilevata in più ho uno shelly EM con la sua relativa pinza amperometrica che mi legge la potenza in ingresso sul differenziale.
Ad ogni presa associo un valore espresso in potenza di due, ovvero quando il valore rilevato dallo shelly EM supera un valore soglia, parte un'automazione che controlla quali di queste quattro prese è accesa  e incrementa il valore della variabile di conseguenza.
Esempio supponendo di inizializzare la variabile "contatore" a 0 e di assegnare 2^0=1 alla lavatrice, 2^1=2 al phon, 2^2=4 al ferro da stiro, 2^3=8 al forno, nel momento stesso in cui la potenza mi supera un determinato valore l'automazione mi controlla quale delle prese è accesa e conseguentemente mi incrementa la variabile "contatore" del relativo valore.
Alla fine avrò che la variabile contatore assumerà un valore tra 0 (tutto staccato ma in questo caso l'automazione non parte) e 15 (tutte le 4 prese collegate). A questo punto sapendo a priori quali prese sono accese posso decidere quale spegnere utilizzando un IF o una Choose  sulla variabile "contatore".
L'alternativa sarebbe fare spegnere tutte quelle che sono accese (magari ad eccetto del forno) e a questo punto non mi serve più la variabile.
Ecco a grandi linee lo scopo dell'utilizzo della variabile.
P.S.: Se non ho inteso male, utilizzando l'input_number l'incremento che posso dare chiamando il Servizio "increment" è fisso e uguale al passo step, Quindi dovrei ad esempio posto step =1 chiamare con un ciclo for il servizio 2, 4 o 8 volte (troppo arzigogolato) per poter immagazzinare nella variabile il valore così come spiegato sopra.

**Reply by 50494554524F** [reply]
se l'automazione parte a che ti serve rifare i "conti"?!? decidi quali prese staccare e basta...

**Reply by Newreader** [reply]
Beh non è proprio così perché potrei decidere di spegnere una presa che in quel momento è già spenta. L'automazione parte e funziona però, non conoscendo quale delle quattro prese in quel momento è accesa,  sono costretto a sondarle tutte una alla volta e se accesa spegnerla. Quindi non ho possibilità di scelta e soprattutto devo intervenire su tutte.

**Reply by 50494554524F** [reply]
puoi definire un group con la lista delle prese da prendere in considerazione... sensori consumo instantaneo o binary_sensor per indicare se attiva o no... se userai un gruppo di binary_sensor potrai usarlo per fa scattare l'automazione

nell'automazione avrai un ciclo su expand(nome_gruppo) e deciderai quale spegnere, quello restituito da expand lo puoi anche ordinare usando i filtri jinja o magari puoi inserirle già ordinate per "carico" quando le aggiungi al gruppo

più pulito...

**Reply by Scialla** [reply]
Se lo scopo è staccare la presa che consuma di più per evitare sovraccarico e distacco dal contatore, io ho fatto automazione utilizzando poi una choose per decidere quale presa staccare in base al consumo istantaneo. Comunque trovi un sacco di package già fatti..

**Reply by Newreader** [reply]
Grazie per gli interessanti suggerimenti.
Per quanto riguarda i package preferisco, in questo caso, impegnarmi personalmente nella soluzione del problema.
Nel mio caso non voglio staccare la presa che consuma di più ma quella in cui è inserito l'elettrodomestico che è meno utile fra quelli, in quel momento, collegati alla rete (esempio preferisco staccare il ferro da stiro piuttosto che il forno anche se quest'ultimo potrebbe nell'occasione consumare di più oppure nel caso dovessero essere attaccate contemporaneamente ferro da stiro, lavatrice e phon posso a priori stabilire di staccare la presa della lavatrice e ferro da stiro). Comunque tutto deve essere automatizzato senza interventi esterni. le scelte vengono operate a monte a secondo delle prese che sono sotto carico.
Per quanto riguarda l'innesco dell'automazione esse è regolato dal superamento di determinato valore prefissato da me della potenza istantanea richiesta dalla rete  e fornitami dallo shelly EM.
Proverò ad adeguare alle mie esigenze i vostri suggerimenti.

**Reply by Scialla** [reply]
Io ho fatto una choose mettendo in fila gli elettrodomestici in ordine di importanza, il più importante in fondo. In sostanza, alo superamento della soglia, controlla nell'ordine stabilito gli assorbimenti dei vari elettrodometsici e, se assorbono troppo, inizia a staccare il primo e ricontrolla il totale, poi passa al secondo, al terzo e così via, finche i consumi non tornano sotto soglia. Basta una choose.
Gli input_number lil uso per cambiare il limite di soglia, il livello di guardia, la soglia istantanea, il tempo di intervento, il massimo consumo consentito per i singoli elettrodomestici prima di staccarli sempre nell'ordine stabilito.

**Reply by Newreader** [reply]
Scusate se non ho risposto prima ma sono stato impegnato. Vorrei nuovamente ringraziarvi per i contributi e le idee preziose che mi avete dato. Ho risolto il problema trovando la soluzione che volevo attraverso l'uso di input_number per gestire il "contatore" e choose per gestire le varie sequenze.
